### PR TITLE
Add base Pane component and migrate RustMirView to TypeScript

### PR DESCRIPTION
**Pull request by supergrecko**
A lot of the code used in the tool panes is repeated because the codebase had to support es5. This patch implements a basic pane container which may be extended by the panes. This should make implementing new panes a lot easier.

To demonstrate usage, the Rust MIR view pane has been migrated to this new Pane component.

For context: I'm proposing this change because I'm in the process of adding more Rust tooling panes. The existing panes may be migrated to use Pane at any time once this is merged.

### CODE REVIEW

File: static/pane.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+ */
+export abstract class Pane<E extends monaco.editor.IEditor> {
+    compilerInfo: PaneCompilerState;
+    container: Container;
+    domRoot: JQuery;
+    topBar: JQuery;
+    eventHub: any /* typeof hub.createEventHub() */;
+    selection: monaco.Selection;
+    editor: E;
+    fontScale: typeof FontScale
```
Comment (supergrecko) [inline static/pane.ts:47]: This is still weakly typed as the entire FontScale module is an es5 class.

File: static/pane.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+     * Initialize the monaco editor instance. Typical implementation for looks
+     * like this:
+     *
+     * ```ts
+     * this.editor = monaco.editor.create(editorRoot, extendConfig({
+     *     // goodies
+     * }));
+     * ```
+     */
+    abstract createEditor(editorRoot: HTMLElement): void;
```
Comment (supergrecko) [inline static/pane.ts:101]: Different panes use different monaco editor kinds and monaco editor options, so delegating the editor creation to a method makes more sense for implementors.

File: static/pane.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+     *
+     * ```ts
+     * ga.proxy('send', {
+     *   hitType: 'event',
+     *   eventCategory: 'OpenViewPane',
+     *   eventAction: 'RustMir',
+     * });
+     * ```
+     */
+    abstract registerOpeningAnalyticsEvent(): void
```
Comment (supergrecko) [inline static/pane.ts:115]: Same idea as createEditor()

File: static/pane.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            this.onCompileResult(id, compiler, result);
+        });
+        this.eventHub.on('compiler', (id, compiler, options, editorId) => {
+            this.onCompiler(id, compiler, options, editorId);
+        });
+        this.eventHub.on('compilerClose', (id) => this.onCompilerClose(id));
+        this.eventHub.on('settingsChange', (settings) => this.onSettingsChange(settings));
+        this.eventHub.on('shown', () => this.resize());
+        this.eventHub.on('resize', () => this.resize());
+    }
```
Comment (supergrecko) [inline static/pane.ts:143]: These events are pretty much standard for each pane, so registering them in a different place reduces code needed for implementors.

File: static/pane.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            compilerName: state.compilerName,
+            editorId: state.editorid,
+        }
+        this.fontScale = new FontScale(this.domRoot, state, this.editor);
+        this.topBar = this.domRoot.find('.top-bar');
+
+        this.registerButtons(state);
+        this.registerStandardCallbacks();
+        this.registerCallbacks();
+        this.setTitle();
```
Comment (RubenRBS) [inline static/pane.ts:78]: Missing a call to `this.registerOpeningAnalyticsEvent()`?

File: static/pane.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2021, Compiler Explorer Authors
```
Comment (RubenRBS) [inline static/pane.ts:1]: I'd move this to `static/panes/pane.ts`

File: static/pane.ts [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        };
+        this.fontScale.addState(state);
+        return state;
+    }
+
+    updateState() {
+        this.container.setState(this.getCurrentState());
+    }
+
+    resize() {
```
Comment (RubenRBS) [inline static/pane.ts:188]: This got changed a few days ago. It now calls a function in `utils.ts` for most panes. Can that be set here, or will we need to override this in almost every pane?

**Review by RubenRBS — APPROVED** [review]

File: static/pane.ts (hunk 5 above)
Comment (supergrecko) [reply static/pane.ts:78]: Yep, thanks!

File: static/settings.interfaces.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+type ColourScheme =
```
Comment (RubenRBS) [inline static/settings.interfaces.ts:25]: Can you add a "Don't forget to edit the settings.interfaces.ts file if you add add/modify a setting!" comment in the settings file? Just as a reminder as long as that file is not ported too :)

File: static/pane.ts (hunk 7 above)
Comment (supergrecko) [reply static/pane.ts]: Done, could you double-check the implementation just to be sure? Thanks!

File: static/panes/pane.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    updateState() {
+        this.container.setState(this.getCurrentState());
+    }
+
+    resize() {
+        const topBarHeight = utils.updateAndCalcTopBarHeight(this.domRoot,
+            this.topBar, this.hideable);
+        this.editor.layout({
```
Comment (RubenRBS) [inline static/panes/pane.ts:192]: Maybe check that editor is not null here, as there could be tools that don't use it

File: static/settings.js [hunk 10]
```diff
@@ -135,6 +135,8 @@ function setupSettings(root, settings, onChange, subLangId) {
         });
     }
 
+    // Don't forget to edit the settings.interfaces.ts file if you add
```
Comment (RubenRBS) [inline static/settings.js:138]: Typo: add add/modify

File: static/tsconfig.json [hunk 11]
```diff
@@ -13,8 +13,11 @@
 		"panes/tree.ts",
 		"alert.ts",
 		"formatter-registry.ts",
+		"panes/pane.ts",
+		"panes/pane.interfaces.ts",
+		"panes/rustmir-view.ts",
 		"sharing.ts",
 		"url.ts",
-		"utils.ts"
+		"utils.ts",
```
Comment (RubenRBS) [inline static/tsconfig.json:21]: Trailing comma in a json file

**Pull request merged**